caput/util/mpitools.py

```diff
@@ -293,293 +309,291 @@ def available_memory_shared() -> int:
+        level_all: int = logging.WARN,
+        comm: MPI.Comm | None = None,
+    ) -> None:
+        super().__init__()
 
+        self.add_mpi_info = add_mpi_info
 
-def allreduce(sendobj, op=None, comm=_comm):
-    """Call comm.allreduce if MPI is enabled."""
-    if comm is not None and comm.size > 1:
-        return comm.allreduce(sendobj, op=(op or MPI.SUM))
+        self.level_rank0 = level_rank0
+        self.level_all = level_all
 
-    return sendobj
+        self.rank = comm.rank if comm else rank
+        self.size = comm.size if comm else size
 
+    def filter(self, record: logging.LogRecord) -> bool:
+        """Return True if we should filter the record."""
+        # Add MPI info if desired
+        try:
+            record.mpi_rank
+        except AttributeError:
+            if self.add_mpi_info:
+                record.mpi_rank = self.rank
+                record.mpi_size = self.size
 
-# def Gatherv(sendbuf, recvbuf, root=0, comm=_comm):
-#     if comm is not None and comm.size > 1:
-#         comm.Gatherv(sendbuf, recvbuf, root=root)
-#     else:
-#         # if they are just numpy data buffer
-#         recvbuf = sendbuf.copy()
-#         # TODO, other cases
+        # Add a new field with the elapsed time in seconds (as a float)
+        record.elapsedTime = record.relativeCreated * 1e-3
 
+        # Return whether we should filter the record or not.
+        return (record.mpi_rank == 0 and record.levelno >= self.level_rank0) or (
+            record.mpi_rank > 0 and record.levelno >= self.level_all
+        )
 
-# def Allgatherv(sendbuf, recvbuf, comm=_comm):
-#     if comm is not None and comm.size > 1:
-#         return _comm.Allgatherv(sendbuf, recvbuf)
-#     else:
-#         # if they are just numpy data buffer
-#         recvbuf = sendbuf.copy()
-#         # TODO, other cases
 
+# Comm Management
+# ===============
 
-def parallel_map(func, glist, root=None, method="con", comm=_comm):
-    """Apply a parallel map using MPI.
 
-    Should be called collectively on the same list. All ranks return the full
-    set of results.
+class _close_message:
+    def __repr__(self) -> str:
+        return "<Close message>"
 
-    Parameters
-    ----------
-    func : function
-        Function to apply.
-    glist : list
-        List of map over. Must be globally defined.
-    root : None or Integer
-        Which process should gather the results, all processes will gather the results if None.
-    method: str
-        How to split `glist` to each process, can be 'con': continuously, 'alt': alternatively, 'rand': randomly. Default is 'con'.
-    comm : MPI communicator
-        MPI communicator that array is distributed over. Default is the gobal _comm.
 
-    Returns
-    -------
-    results : list
-        Global list of results.
-    """
-    # Synchronize
-    barrier(comm=comm)
+def active_comm(aprocs: list) -> MPI.Comm | None:
+    """Return a communicator consists of a list of processes in `aprocs`."""
```

Comment:
```suggestion
    """Return a communicator consisting of processes listed in `aprocs`."""
```

(Also, probably all of these functions that take and `aprocs` should have a Parameters section.)